Academy/AI/1-Intro-To-ML-And-MLOps/lab.ipynb

```diff
@@ -0,0 +1,563 @@
+    "- **Metrics** → Performance indicators of the model, such as **accuracy**, **F1-score**, precision, and recall.  \n",
+    "- **Model** → The trained model version, which can be reloaded and reused without retraining.  \n",
+    "\n",
+    "By logging these elements, we can analyze and compare different model versions in a structured and reproducible way."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Use the same type of model used in Part 1\n",
+    "import os\n",
+    "from sklearn.ensemble import RandomForestClassifier\n",
+    "from sklearn.metrics import accuracy_score, f1_score\n",
+    "from mlflow.models.signature import infer_signature\n",
+    "\n",
+    "# Execute 4 experiments to train multiple models \n",
+    "n_estimators = [1, 10, 100, 500]\n",
+    "\n",
+    "for n_e in n_estimators:\n",
+    "    # Create a new MLFlow run\n",
+    "    with mlflow.start_run():\n",
+    "        # Log Param\n",
+    "        mlflow.log_param(\"n_estimators\", n_e)\n",
+    "\n",
+    "        # Create and train the model instance\n",
+    "        rf_model = RandomForestClassifier(n_estimators=n_e, random_state=42)\n",
+    "        rf_model.fit(X_train, y_train)\n",
+    "\n",
+    "        # Compute metrics\n",
+    "        y_pred = rf_model.predict(X_test)\n",
+    "        accuracy = accuracy_score(y_test, y_pred)\n",
+    "        f1 = f1_score(y_test, y_pred, average='weighted')\n",
+    "        mlflow.log_metric(\"accuracy\", accuracy)\n",
+    "        mlflow.log_metric(\"f1\", f1)\n",
+    "\n",
+    "        # Create heatmap with Seaborn\n",
+    "        target_names = [\"No Rain\", \"Rain\"]\n",
+    "        cm = confusion_matrix(y_test, y_pred)\n",
+    "        plt.figure(figsize=(6, 4))\n",
+    "        sns.heatmap(cm, annot=True, fmt=\"d\", cmap=\"Blues\", xticklabels=target_names, yticklabels=target_names)\n",
+    "\n",
+    "        # Add titles\n",
+    "        plt.xlabel(\"Predicted Label\")\n",
+    "        plt.ylabel(\"True Label\")\n",
+    "        plt.title(\"Confusion Matrix\")\n",
+    "\n",
+    "        # Save the plot as PNG\n",
+    "        if not os.path.exists(\"dev/\"):\n",
+    "            os.makedirs(\"dev/\")\n",
+    "        plt.savefig(\"dev/confusion_matrix.png\")\n",
+    "        plt.close()\n",
+    "        # Save the confusion matrix on MLFlow as artifact\n",
+    "        mlflow.log_artifact(\"dev/confusion_matrix.png\")\n",
+    "\n",
+    "        # Save the model on MLFlow\n",
+    "        example_dict = {'MinTemp': 1.1, 'MaxTemp': 1.1, 'Humidity3pm': 1.1, 'Humidity9am': 1.1}\n",
+    "        signature = infer_signature(model_input=example_dict)\n",
+    "        mlflow.sklearn.log_model(rf_model, \"random_forest_model\", signature=signature)\n",
+    "\n",
+    "        print(f\"Experiment finished. Registered accuracy: {accuracy:.2f}\")"
+   ]
+  },
```

Comment:
While trying to reproduce this on a standard Debian 12, within a Python Virtual Environment with all the requirements installed, I'm getting this:

```
Traceback (most recent call last):
  File "/home/rasca/Labs/venvs/aisetup/mlflow.py", line 3, in <module>
    import mlflow
  File "/home/rasca/Labs/venvs/aisetup/mlflow.py", line 4, in <module>
    import mlflow.sklearn
ModuleNotFoundError: No module named 'mlflow.sklearn'; 'mlflow' is not a package
```

Maybe it depends on the specific package versions used while writing the lab, so having a requirements.txt would help, but maybe it depends on something else.

Since Python venv should prevent these kind of errors, something needs to be changed/integrated here.

Reply:
@rndg  just to confirm that I've tried again from scratch, using exactly the steps mentioned in the doc and the result is still the mentioned error.